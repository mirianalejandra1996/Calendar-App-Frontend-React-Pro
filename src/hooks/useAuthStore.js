import { useDispatch, useSelector } from "react-redux";
import calendarApi from "../api/calendarApi";
import { clearErrorMessage, onChecking, onLogin, onLogout, onLogoutCalendar } from "../store";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({email, password}) => {
        
    console.log({email, password})
        dispatch(onChecking())
        try {
            const { data } = await calendarApi.post('/auth', { email, password })
            localStorage.setItem('token', data.token)
            localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
            dispatch(onLogin({name: data.name, uid: data.uid}))
        } catch (error) {
            console.log('error', error.message)
            localStorage.removeItem('token')
            localStorage.removeItem('token-init-date')
            dispatch(onLogout('Invalid credentials'))
            setTimeout(() => dispatch(clearErrorMessage()) , 10)
        }
  }

  const startRegister = async ({ name, email, password }) => {
    
    try {
      const { data } = await calendarApi.post('/auth/new', { name, email, password})


      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
      // with the time
      dispatch(onLogin({name: data.name, uid: data.uid}))

    } catch (error) {
      console.log('error', error)

      // const {error: errorFieds} = error.response.data 
      const { errors, msg } =  error.response.data 

      if (errors) dispatch(onLogout('Please fill all fields correctly'))
      // else if (msg === 'Email not available')  dispatch(onLogout(msg))
      else if (msg)  dispatch(onLogout(msg === 'Email not available' ? 'Email already in use' : msg))

      else dispatch(onLogout(error.message))

      setTimeout(() => dispatch(clearErrorMessage()) , 10)
    }
  }

  const checkAuthToken =  async() => {
    
    const token = localStorage.getItem('token')
    if  (!token) return dispatch(onLogout())

    try {

        const { data } = await calendarApi.get('/auth/renew')

        // * Note: I can create a functionality to set cookies for code optimization
        localStorage.setItem('token', data.token)
        localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
        // with the time

        dispatch(onLogin({name: data.name, uid: data.uid}))


      } catch (error) {
            console.log('error', error)
            localStorage.clear()
            dispatch(onLogout())
    }
  }

  const startLogout = () => {
    dispatch(onLogout())
    dispatch(onLogoutCalendar())
    localStorage.clear()
  }

  return {
    // * Properties
    status,
    user,
    errorMessage,

    // * Methods
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister,
  };
};
