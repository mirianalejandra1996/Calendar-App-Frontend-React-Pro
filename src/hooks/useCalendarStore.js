import { useDispatch, useSelector } from "react-redux";
import {
  onAddNewEvent,
  onLogoutCalendar,
  onDeleteEvent,
  onLoadEvents,
  onSetActiveEvent,
  onUpdateEvent,
} from "../store/calendar/calendarSlice";
import calendarApi from "../api/calendarApi";
import { convertEventsToDateEvents } from "../helpers";
import Swal from "sweetalert2";

export const useCalendarStore = () => {
  const dispatch = useDispatch();

  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector((state) => state.auth);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  // No vamos a usar un thunks asíncronos, de frente vamos a disparar acciones síncronas
  const startSavingEvent = async (calendarEvent) => {

    try {
      if (calendarEvent.id) {
        // If user has an ID is because the event needs to be updated
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
        dispatch(onUpdateEvent({ ...calendarEvent }));
        Swal.fire('Update', "It has been updated successfully", 'success');
        return;
      } else {
        // Else is creating a new event and adding an unique ID event
        const { data } = await calendarApi.post("/events", calendarEvent);
        dispatch(onAddNewEvent({ ...calendarEvent, id: data.event.id, user: { ...user, _id: user.uid} }));
        Swal.fire('Creation', "It has been created successfully", 'success');
      
      }
    } catch (error) {
      console.log('error', error)
      Swal.fire('Error while saving event', error.response.data?.msg , 'error' )
    }
  };

  const startDeletingEvent = async () => {

    try {
      await calendarApi.delete(`/events/${ activeEvent.id }` );
      dispatch(onDeleteEvent());
      Swal.fire('Elimination', "It has been removed successfully", 'success');
      return;
      
    } catch (error) {
      console.log('error', error)
      Swal.fire('Error while deleting event', error.response.data?.msg , 'error' )
    }
  };

  const startLoadingEvents = async () => {

    try {
      const { data } = await calendarApi.get("/events");

      console.log('loading events', data)
      const events = convertEventsToDateEvents(data.events);

      dispatch(onLoadEvents(events));
    } catch (error) {
      console.log("Error loading events", error);
    }
  };  

  return {
    // * Propiedades
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,

    // * Métodos
    setActiveEvent,
    startDeletingEvent,
    startSavingEvent,
    startLoadingEvents,
  };
};
