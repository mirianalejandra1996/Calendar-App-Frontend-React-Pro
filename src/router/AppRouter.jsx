import { Navigate, Route, Routes } from "react-router-dom"
import { LoginPage } from "../auth"
import { CalendarPage } from "../calendar"
import { useAuthStore } from "../hooks"
import { useEffect } from "react"

import { LoadingPage } from "../shared"





export const AppRouter = () => {

  const { status, checkAuthToken} = useAuthStore()

  // console.log('env', process.env.NODE_ENV)
  console.log('env', process.env)

    useEffect(() => {
      checkAuthToken()
    }, [])

    if (status === 'checking') {
        return <LoadingPage/>
    }
    

  return (
    <Routes>

        {
            (status === 'not-authenticated')
            ? (
              <>
                <Route path="auth/*" element={<LoginPage/>}/>
                <Route path="/*" element={<Navigate to={'/auth/login'}/>}/>

              </>
            )
            : (
              <>
                <Route path="/" element={<CalendarPage/>}/>
                <Route path="/*" element={<Navigate to="/" />}/>
              </>
            )
        }

    </Routes>
  )
}