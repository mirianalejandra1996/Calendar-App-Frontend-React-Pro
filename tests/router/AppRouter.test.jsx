// import { useCalendarStore } from "../../src/hooks/useCalendarStore";
import { useAuthStore } from "../../src/hooks/useAuthStore";
// import { useUiStore } from "../../src/hooks/useUiStore";

import { render, screen } from "@testing-library/react";

import { AppRouter } from "../../src/router/AppRouter";

import { MemoryRouter } from "react-router-dom"

jest.mock("../../src/hooks/useAuthStore")


// En lugar de renderizar TODO los elementos que contienen ese componente, se estará
// reemplazando por un h1 solo para probar que llega a ver la pantalla del calendario.
jest.mock("../../src/calendar", () => ({
    CalendarPage: () => <h1>CalendarPage</h1>
}))

describe('Tests on <AppRouter />', () => {

    const mockCheckAuthToken = jest.fn()

    beforeEach(() => jest.clearAllMocks() )


    test('should display loading screen and call checkAuthToken fn', () => {
        // const { status, checkAuthToken} = useAuthStore()

        useAuthStore.mockReturnValue({
            status : 'checking',
            checkAuthToken: mockCheckAuthToken,
        })

        render(<AppRouter/>)

        // screen.debug()

        expect(screen.getByLabelText("loading-spinner")).toBeTruthy()

        expect(mockCheckAuthToken).toHaveBeenCalled()
        expect(mockCheckAuthToken).toHaveBeenCalledTimes(1)
    })

    test('should display login screen if user is not logged in', () => {

        useAuthStore.mockReturnValue({
            status : 'not-authenticated',
            checkAuthToken: mockCheckAuthToken,
        })

        const { container } = render(
            // Estoy en una ruta diferente al login para confirmar si me envía a /auth/login
            // el navigate funciona!
            <MemoryRouter initialEntries={[`/`]}>
                <AppRouter />
            </MemoryRouter>
        )

        // screen.debug()

        expect(screen.getByText('Log In')).toBeTruthy()
        
        // Porque considero que esta pantalla no será modificada luego.
        expect(container).toMatchSnapshot()
    })

    test('should display calendar screen if user is logged in', () => {

        useAuthStore.mockReturnValue({
            status : 'authenticated',
            checkAuthToken: mockCheckAuthToken,
        })
        
        render(
            // Estoy en la ruta del login para confirmar si me envía a la ruta "/"
            // el navigate funciona!
                <MemoryRouter initialEntries={[`/auth/login`]}>
                    <AppRouter />
                </MemoryRouter>
        )


        // screen.debug()

        // Hubiera tenido un problema ya que tendría que mockear los otros custom hooks que
        // están involucrados en el CalendarPage, asi que lo que vamos a hacer es mockear el CalendarPage.
        // * Ver arriba el mock


        expect(screen.getByText('CalendarPage')).toBeTruthy()
        
    })
})