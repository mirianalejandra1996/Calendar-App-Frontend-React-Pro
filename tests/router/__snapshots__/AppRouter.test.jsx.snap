// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <AppRouter /> should display login screen if user is not logged in 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Log In
        </h3>
        <form
          aria-label="login-form"
        >
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="loginEmail"
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="loginPassword"
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Sign Up
        </h3>
        <form
          aria-label="register-form"
        >
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="registerName"
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="registerEmail"
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="registerPassword"
              class="form-control"
              name="registerPassword"
              placeholder="ContraseÃ±a"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="registerPassword2"
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
